### ISSUE TITLE
Invalid action on dispatch

### ISSUE THREAD

**Issue by jonystorm** [kraihn/cordova-plugin-tag-manager#7]
Hi @kraihn ,

The error calback of the dispatch function triggers and it's returning an "Invalid action" msg.

Any suggestion what does it mean? or why I'm getting this error?

Is there any other plugin that may be needed like com.google.playsevices or anything to solve this?

**Comment by kraihn** [comment]
Can you attach the JS calls you make to get this error? No other plugins should be required. It's was intentional to not require play services. Also, can you tell me if your TagManager container is V1 or V2?

**Comment by jonystorm** [comment]
I'm using TagManager V2. Could this be the problem?

This is the piece of code I use:

var tagManager = cordova.require('com.jareddickson.cordova.tag-manager.TagManager'),
        trackingId = window.device.platform === 'Android' ? 'My-GTM-Code' : 'Ios Code',
        intervalPeriod = 30;

```
var success = function(callback) {
    console.log('SUCCESS_GTM');
    console.log(callback);
};
var error = function(callback) {
    console.log('ERROR_GTM');
    console.log(callback);
};
var errorDis = function(callback) {
    console.log('ERROR_DISPATCH_GTM');
    console.log(callback);
};

// Initialize Tag Manager
tagManager.init(success, error, trackingId, intervalPeriod);
tagManager.trackEvent(success, error, 'Init', 'Any-Event', 'Any-Event', 1);
tagManager.dispatch(success,errorDis);
```

The dispatch function triggers errorDis function. And an "Invalid action" msg from console is also shown on the console.log(callback).

All of this code is executed after deviceready.

**Comment by kraihn** [comment]
I'll have to check into this later, but the version could be the problem. All development has been against V1, and we're a little behind in getting ready for the V2 migration.

**Comment by jonystorm** [comment]
I found this message on the JS log: 

GoogleTagManager: No asset file: tagmanager/GTM-XXXX found.
GoogleTagManager: No asset file: tagmanager/GTM-XXXX.json found (or errors reading it).
GoogleTagManager: No default container found; creating an empty container.

Is it ok this way? or do I have to add the binary file that provides GTM when creating a container?

I'm trying with a GTM V1 container, but this is what I still get:

05-13 22:10:21.487  3934  3934 D CordovaLog: file:///android_asset/www/js/main.js: Line 50 : initGTM - id = GTM-XXXXXX; interval = 30 seconds
05-13 22:10:22.337  3934  3934 W CordovaPlugin: Attempted to send a second callback for ID: TagManagerXXXXXX
05-13 22:10:22.337  3934  3934 W CordovaPlugin: Result was: "Invalid action"
05-13 22:10:22.337  3934  3934 D CordovaLog: file:///android_asset/www/js/main.js: Line 53 : ERROR_GTM
05-13 22:10:22.337  3934  3934 D CordovaLog: file:///android_asset/www/js/main.js: Line 54 : trackEvent failed - not initialized
05-13 22:10:23.387  3934  3934 D CordovaLog: file:///android_asset/www/js/main.js: Line 57 : ERROR_DISPATCH_GTM
05-13 22:10:23.387  3934  3934 D CordovaLog: file:///android_asset/www/js/main.js: Line 58 : Invalid action

Don't know if is any help, but these are all the msgs I found.

**Comment by jonystorm** [comment]
I tracked the "Invalid action" error to the CDVTagManager.java to this piece of code:

if (inited) {
                try {
                    GAServiceManager.getInstance().dispatchLocalHits();
                }
                catch (final Exception e) {
                    callback.error(e.getMessage());
                }
            }
            else {
                callback.error("dispatch failed - not initialized");
            }

The "Invalid action" msg comes from the failed try of the function GAServiceManager.getInstance().dispatchLocalHits(); I assume it's the error msg from the GA API, so I come back to the idea the error might be those missing files at the top of my previous comment.

Is it possible?

**Comment by kraihn** [comment]
It sounds to me like the initialization failed. The way I'm using it now there is no default container stored with the app. It gets pulled in from the web. Is your GTM Id configured correctly? A dispatch is going to keep failing until the initialization completes.

**Comment by jonystorm** [comment]
I made some more tests, and everything regarding the init seems fine. I tried to do the same but with a blank id, and the trackpage function return an error with msg "not initialized", that doesn't appear when using my GTM id. Although, the dispatch function throws the same error ("invalid action") when msg "not initialized" it's shown by the trackpage function.

**Comment by kraihn** [comment]
Thank you for being so helpful with your feedback. I have allotted time over the next two weeks to help resolve this issue, and most likely to add support for Tag Manager V2.

**Comment by jonystorm** [comment]
Great! I'm looking forward to it! 
I'm finishing an app I'd really like to have GTM V2 working on it, I wasn't able to find any other plugin that works properly.

**Comment by kraihn** [comment]
I hit the "Invalid action" error this morning. Not sure the cause yet, but I'm looking into it. The interesting thing is that the automatic dispatch is still happening. I'm still seeing events and pageviews being sent to my GA account through Tag Manager.

**Comment by kraihn** [comment]
Hey @jonystorm, I think I have a fix for the issue you're seeing. Can you take a look at this PR?

https://github.com/kraihn/cordova-plugin-tag-manager/pull/9

**Comment by jonystorm** [comment]
Hi @kraihn , I'm testing it right now. Now the console is showing the success callbacks being call correctly. However, I'm still not getting any hits on GA. 

I tried using a GA plugin to test and it records events. I know GA hits may take up to a few hours to show up, though with GA plugin showed up right away, but I'll keep checking it and testing.

**Comment by kraihn** [comment]
Hmm, I feel a little stuck on this. I've tested the application in both Android and iOS, and for my end it's working fine. Yesterday I did a test migration from Tag Manager V1 to Tag Manager V2 and saw pageview and event hits for both. Is your Tag Manager container setup correctly? That would be my first guess. The steps to get Tag Manager to forward hits to GA is a manual step that requires creating 6 Macros, 2 Rules, and 2 Tags.

**Comment by jonystorm** [comment]
But those steps are for GTM V1. On V2 you only have to add the GA label, add the GA id, and set the trigger event (for which I use "Any event").

Is it possible that maybe cordova version has anything to do with it? I'm using version 5.0.0 

I know we are close to figuring it out, probably I'm missing something.
Is there any chance that you upload a working copy of a sample cordova project for me to check what may be happening?

**Comment by kraihn** [comment]
I don't believe that is the case. You should still manually create the Tags, Triggers and Variables. For all of my testing, I have set the mapping of values sent from the mobile app to what GTM passes on to GA. It's unlikely that the version of Cordova being used will have any effect. Please try the manual setup steps in GTM and let me know if it starts working.

In the meantime, I'll see about putting together a sample project.

**Comment by jonystorm** [comment]
Hi @kraihn , I've been testing different configurations on GTM V2, and still I am not getting any hits on GA. 

I don't know what else to do, on the JS console side, everything seems to be fine now, no errors at all.

Have you tried with a fresh new GTM V2 container (and not with a migrated one)?
I know this probably isn't it, but it's the only thing I found we did differently. I don't know what else to test.

**Comment by jonystorm** [comment]
One thing that I didn't understand was that on GTM V2, the built-in variable, Event, represents the string containing the value stored in the ‘event’ dataLayer key. 
And for what I found, what is being send is a DataLayer.mapOf("event", "eventAction",...)

My question would be, where is GTM receiving (in which variable) the category, eventAction and eventLabel? (from function trackEvent) 

Maybe this is where I'm missing out something about GTM V2 configuration, I started using GTM recently straight with version 2.

Thanks,
Jonatan
